incubator/hnc/internal/reconcilers/hnc_config.go

```diff
@@ -98,6 +110,92 @@ func (r *ConfigReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	return ctrl.Result{}, syncErr
 }
 
+// reconcileTypes makes sure there's no dup and the types exist. Update the type
+// set with GR to GVK mappings.
+func (r *ConfigReconciler) reconcileTypes(inst *api.HNCConfiguration) error {
+	// Get all resources for all groups.
+	allRes, err := r.getAllResources()
+	if err != nil {
+		r.Log.Error(err, "while trying to get all resources")
+		return err
+	}
+	// exist keeps track of existing types in the `config` singleton.
+	exist := grSet{}
+	r.activeGRs = grSet{}
+	r.activeGVKs = gvkSet{}
+	for _, t := range inst.Spec.Types {
+		// If there are multiple configurations of the same type, we will follow the first
+		// configuration and ignore the rest.
+		if !r.ensureNoDuplicateTypeConfigurations(inst, t, exist) {
+			continue
+		}
+		// Look if the resource exists in the API server.
+		gr := schema.GroupResource{Group: t.Group, Resource: t.Resource}
+		if err := r.addType(gr, allRes); err != nil {
+			// If the type is not found, log error and write conditions but don't
+			// early exit since the other types can still be reconciled.
+			r.Log.Error(err, "while trying to get the resource")
+			r.writeCondition(inst, api.TypeNotFound, err.Error())
+		}
+	}
+	return nil
+}
+
+func (r *ConfigReconciler) getAllResources() ([]*restmapper.APIGroupResources, error) {
+	// Get all resources for all groups.
+	dc, err := discovery.NewDiscoveryClientForConfig(r.Manager.GetConfig())
+	if err != nil {
+		r.Log.Error(err, "while creating discovery client")
+		return nil, err
+	}
+	return restmapper.GetAPIGroupResources(dc)
+}
+
+// ensureNoDuplicateTypeConfigurations checks whether the configuration of a
```

Comment:
Any reason you moved this function around? It would be easier to compare if you left it where it was (just before syncRemovedReconcilers)

Reply:
Ah it's in the order of being called after the caller function. Since it's now called when traversing all types at the beginning, instead of being called in `syncActiveReconcilers` after the active types are already got.

Reply:
Yeah, makes sense. When I'm making a big change like this, I generally try to keep things in the same order to make it easy to see what's changed in the diff. Then you can do a separate "rearrange the functions" cleanup PR later. Up to you if you want to change it though.

Reply:
Removed this func and inlined it with the factored out `GVKFor()` func.